### PR TITLE
feat: the request parameter participating in the signature can be the parameter name

### LINKED ISSUE #2823: feat(hmac-auth):  the request parameter participating in the signature can be the parameter name
### Issue description

When the request parameters of the hmac-auth plug-in participate in the signature. If the parameter is in the form of `key` instead of `key=value` or `key=`, the Lua thread will be aborted and the signature will fail. Therefore, we need to be compatible with this `key` situation.

Example：

Add a test case in `hmac-auth.t`, contains request parameters in the form of `key`(name=jack&age):

```
=== TEST 41: verify: ok
--- config
location /t {
    content_by_lua_block {
        local ngx_time   = ngx.time
        local ngx_http_time = ngx.http_time
        local core = require("apisix.core")
        local t = require("lib.test_admin")
        local hmac = require("resty.hmac")
        local ngx_encode_base64 = ngx.encode_base64

        local secret_key = "my-secret-key6"
        local timestamp = ngx_time()
        local gmt = ngx_http_time(timestamp)
        local access_key = "my-access-key6"
        local custom_header_a = "asld$%dfasf"
        local custom_header_b = "23879fmsldfk"

        local signing_string = {
            "GET",
            "/hello",
            "age&name=jack",
            access_key,
            gmt,
            "x-custom-header-a:" .. custom_header_a,
            "x-custom-header-b:" .. custom_header_b
        }
        signing_string = core.table.concat(signing_string, "\n") .. "\n"
        core.log.info("signing_string:", signing_string)

        local signature = hmac:new(secret_key, hmac.ALGOS.SHA256):final(signing_string)
        core.log.info("signature:", ngx_encode_base64(signature))
        local headers = {}
        headers["X-HMAC-SIGNATURE"] = ngx_encode_base64(signature)
        headers["X-HMAC-ALGORITHM"] = "hmac-sha256"
        headers["Date"] = gmt
        headers["X-HMAC-ACCESS-KEY"] = access_key
        headers["X-HMAC-SIGNED-HEADERS"] = "x-custom-header-a;x-custom-header-b"
        headers["x-custom-header-a"] = custom_header_a
        headers["x-custom-header-b"] = custom_header_b

        local code, body = t.test('/hello?name=jack&age',
            ngx.HTTP_GET,
            "",
            nil,
            headers
        )

        ngx.status = code
        ngx.say(body)
    }
}
--- request
GET /t
--- response_body
passed
--- no_error_log
[error]
```

Test case run results：

截屏2020-11-23 下午1 59 39

### Environment

* apisix version (cmd: `apisix version`): 2.0
* OS:

### PR DESCRIPTION
**Pull request by Firstsawyou**
### What this PR does / why we need it:

close #2823

### Pre-submission checklist:

* [x] Did you explain what problem does this PR solve? Or what new features have been added?
* [x] Have you added corresponding test cases?
* [ ] Have you modified the corresponding document?
* [x] Is this PR backward compatible?

### CODE REVIEW

File: apisix/plugins/hmac-auth.lua [hunk 1]
```diff
@@ -224,6 +224,13 @@ local function generate_signature(ctx, secret_key, params)
 
         for _, key in pairs(keys) do
             local param = args[key]
+            -- when args without `=<value>`, value is treated as true.
+            -- In order to be compatible with args lacking `=<value>`,
+            -- we need to replace true with an empty string.
+            if param == true then
```
Comment (membphis) [inline apisix/plugins/hmac-auth.lua:230]: `type(param) == "boolean"` is better

**Review by membphis — COMMENTED** [review]
LGTM, except minor code style

File: apisix/plugins/hmac-auth.lua (hunk 1 above)
Comment (Firstsawyou) [reply apisix/plugins/hmac-auth.lua]: Nice, updated.

File: apisix/plugins/hmac-auth.lua [hunk 2]
```diff
@@ -224,6 +224,13 @@ local function generate_signature(ctx, secret_key, params)
 
         for _, key in pairs(keys) do
             local param = args[key]
+            -- when args without `=<value>`, value is treated as true.
+            -- In order to be compatible with args lacking `=<value>`,
+            -- we need to replace true with an empty string.
+            if type(param) == "boolean" then
+                param = ""
```
Comment (tokers) [inline apisix/plugins/hmac-auth.lua:231]: Have you investigated implementations in other languages like `node.js`, `golang`, `python`?

Comment (Firstsawyou) [reply apisix/plugins/hmac-auth.lua:231]: I tested it, and the following is my test result：

node.js: curl "http://localhost:8080/args?name=jack&age"

```
node.js, args name：jack
node.js, args age：
```

python: curl "http://localhost:8080/args?name=jack&age"

```
ParseResult(scheme='http', netloc='127.0.0.1:8080', path='/args', params='', query='name=jack&age', fragment='')
{'age': [''], 'name': ['jack']}
```

golang: curl  "http://localhost:8080/args?name=jack&age"

```
map[age:[] name:[jack]]
args value:  jack 
```

File: t/plugin/hmac-auth.t [hunk 3]
```diff
@@ -1633,3 +1633,125 @@ GET /t
...
 --- no_error_log
 [error]
+
+
+
+=== TEST 42: verify: ok, the request parameter is missing `=<value>`.
+--- config
+location /t {
+    content_by_lua_block {
+        local ngx_time   = ngx.time
```
Comment (tokers) [inline t/plugin/hmac-auth.t:1643]: Style: too many spaces before `=`.

File: t/plugin/hmac-auth.t [hunk 4]
```diff
@@ -1633,3 +1633,125 @@ GET /t
...
+            "GET",
+            "/hello",
+            "age=&name=jack",
+            access_key,
+            gmt,
+            "x-custom-header-a:" .. custom_header_a,
+            "x-custom-header-b:" .. custom_header_b
+        }
+        signing_string = core.table.concat(signing_string, "\n") .. "\n"
+        core.log.info("signing_string:", signing_string)
```
Comment (tokers) [inline t/plugin/hmac-auth.t:1667]: It this your debug logging? You don't check it by `grep_error_log`.

File: t/plugin/hmac-auth.t [hunk 5]
```diff
@@ -1633,3 +1633,125 @@ GET /t
...
+--- no_error_log
+[error]
+
+
+
+=== TEST 43: verify: ok, the value of the request parameter is true.
+--- config
+location /t {
+    content_by_lua_block {
+        local ngx_time   = ngx.time
```
Comment (tokers) [inline t/plugin/hmac-auth.t:1704]: Ditto, too many spaces before `=`.

File: t/plugin/hmac-auth.t [hunk 6]
```diff
@@ -1633,3 +1633,125 @@ GET /t
...
+--- request
+GET /t
+--- response_body
+passed
+--- no_error_log
+[error]
+
+
+
+=== TEST 43: verify: ok, the value of the request parameter is true.
```
Comment (tokers) [inline t/plugin/hmac-auth.t:1700]: Why checking it when value is literal `true` string? It makes no sense for your fixing, the process of literal string `true` is same with other non empty values, you should test arguments with form of "id=123&name".

File: t/plugin/hmac-auth.t (hunk 3 above)
Comment (Firstsawyou) [reply t/plugin/hmac-auth.t]: Good catch.

File: t/plugin/hmac-auth.t (hunk 4 above)
Comment (Firstsawyou) [reply t/plugin/hmac-auth.t:1667]: Yes, this is the debug log, the test does not need to care about it.

File: t/plugin/hmac-auth.t (hunk 5 above)
Comment (Firstsawyou) [reply t/plugin/hmac-auth.t]: done.

File: t/plugin/hmac-auth.t (hunk 6 above)
Comment (Firstsawyou) [reply t/plugin/hmac-auth.t:1700]: Yes, this test is to cover requests whose request parameter value is `true`. The situation you mentioned has been covered in the previous test case(`TEST 42`). ^ _ ^

Comment (tokers) [reply t/plugin/hmac-auth.t:1700]: OK, i see.

File: t/plugin/hmac-auth.t (hunk 4 above)
Comment (tokers) [reply t/plugin/hmac-auth.t:1667]: OK.

**Review by tokers — APPROVED** [review]

**Comment by tokers** [comment]
LGTM.

**Review by membphis — APPROVED** [review]

**Pull request merged**